File: python/open_guardrail/guards/pii_redact_consistency.py
"""Checks that redaction markers and raw PII do not coexist."""

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_REDACTION_MARKERS = re.compile(r"\[(EMAIL|PHONE|REDACTED)\]")

_RAW_PII = [
    ("email", re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")),
    ("ssn", re.compile(r"\b\d{3}-\d{2}-\d{4}\b")),
    ("phone", re.compile(r"\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b")),
]


class _PiiRedactConsistency:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "pii-redact-consistency"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        has_markers = bool(_REDACTION_MARKERS.search(text))
        leaked: List[str] = []
        if has_markers:
            for label, p in _RAW_PII:
                if p.search(text):
                    leaked.append(label)
        triggered = len(leaked) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="pii-redact-consistency",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Raw PII coexists with redaction markers: {', '.join(leaked)}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"leaked_types": leaked} if triggered else None,
        )


def pii_redact_consistency(*, action: str = "block") -> _PiiRedactConsistency:
    return _PiiRedactConsistency(action=action)

File: packages/guards/src/pii-redact-consistency.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PiiRedactConsistencyOptions {
  action: 'block' | 'warn';
  /** Redaction markers to expect */
  markers?: string[];
}

const DEFAULT_MARKERS = ['[EMAIL]', '[PHONE]', '[SSN]', '[NAME]', '[ADDRESS]', '[PII]', '[REDACTED]', '[MASKED]'];

const PII_PATTERNS = [
  /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/,
  /\b\d{3}-\d{2}-\d{4}\b/,
  /\b(?:\+?\d{1,3}[-.\s]?)?\(?\d{2,4}\)?[-.\s]?\d{3,4}[-.\s]?\d{4}\b/,
];

export function piiRedactConsistency(options: PiiRedactConsistencyOptions): Guard {
  const markers = options.markers ?? DEFAULT_MARKERS;

  return {
    name: 'pii-redact-consistency',
    version: '0.1.0',
    description: 'Ensures PII was consistently redacted — no raw PII alongside redaction markers',
    category: 'privacy',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const hasMarkers = markers.some((m) => text.includes(m));
      if (!hasMarkers) {
        return { guardName: 'pii-redact-consistency', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      }
      const leaks: string[] = [];
      for (const pat of PII_PATTERNS) {
        const m = text.match(pat);
        if (m) leaks.push(m[0].slice(0, 10) + '***');
      }
      const triggered = leaks.length > 0;
      return {
        guardName: 'pii-redact-consistency', passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { leaks, message: 'Raw PII found alongside redaction markers' } : undefined,
      };
    },
  };
}
